src/SignalR/clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -501,3 +502,70 @@ export class HttpConnection implements IConnection {
 function transportMatches(requestedTransport: HttpTransportType | undefined, actualTransport: HttpTransportType) {
     return !requestedTransport || ((actualTransport & requestedTransport) !== 0);
 }
+
+export class TransportSendQueue {
+    private sending: boolean = false;
+    private buffer: any[] = [];
+    private promiseQueue?: Promise<void>;
+
+    public async send(transport: ITransport, data: string | ArrayBuffer): Promise<void> {
+        return this.sending ?
+            this.bufferData(transport, data) :
+            this.sendCore(transport, data);
+    }
+
+    private async sendCore(transport: ITransport, data: string | ArrayBuffer) {
+        this.sending = true;
+
+        let resolve: () => void;
+        this.promiseQueue = new Promise((r) => resolve = r);
+
+        try {
+            await transport.send(data);
+        } finally {
+            this.sending = false;
+            this.promiseQueue = undefined;
+            resolve!();
+        }
+    }
+
+    private bufferData(transport: ITransport, data: string | ArrayBuffer) {
+        if (this.buffer.length) {
+            if (typeof(this.buffer[0]) !== typeof(data)) {
+                throw new Error(`Expected data to be of type ${typeof(this.buffer)} but was of type ${typeof(data)}`);
+            }
+        } else {
+            // For the first entry, update the promise to queue up a buffer send
+            this.promiseQueue = this.promiseQueue!
+                .then(() => this.sendBufferedData(transport));
```

Comment:
I think this would be easier to follow if this was written as an async loop instead of using `.then()`.